File: config/manager_secret/rust/src/errors/secret_error.rs
//! Why a secret could not be resolved. Never a default, never an empty value.

use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SecretError {
    /// The component asked for a name it did not declare.
    Undeclared { name: String, declared: Vec<String> },
    /// The provider has no entry under that name.
    Unresolvable { name: String, provider: String },
    /// The provider itself failed -- unreadable mount, unreachable store.
    Provider { name: String, provider: String, detail: String },
}

impl fmt::Display for SecretError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Undeclared { name, declared } => write!(
                f,
                "secret {name:?} was requested but is not declared by this component. Declared: \
                 [{}]. Add it to the component's secret manifest, or stop requesting it -- a \
                 provider that answered undeclared names would give every component the whole \
                 store.",
                declared.join(", ")
            ),
            Self::Unresolvable { name, provider } => write!(
                f,
                "secret {name:?} is declared but the {provider} provider has no entry for it. \
                 There is no default and no empty fallback: a component that continued here would \
                 authenticate with a blank credential."
            ),
            Self::Provider { name, provider, detail } => {
                write!(f, "the {provider} provider failed resolving {name:?}: {detail}")
            }
        }
    }
}

impl std::error::Error for SecretError {}

File: config/manager_secret/elixir/lib/serviceradar_secret/secret_error.ex
defmodule ServiceradarSecret.SecretError do
  @moduledoc "Why a secret could not be resolved. Never a default, never an empty value."

  @spec message(term()) :: String.t()
  def message({:undeclared, name, declared}) do
    "secret #{inspect(name)} was requested but is not declared by this component. " <>
      "Declared: [#{Enum.join(declared, ", ")}]. Add it to the component's secret manifest, or " <>
      "stop requesting it -- a provider that answered undeclared names would give every " <>
      "component the whole store."
  end

  def message({:unresolvable, name, provider}) do
    "secret #{inspect(name)} is declared but the #{provider} provider has no entry for it. " <>
      "There is no default and no empty fallback: a component that continued here would " <>
      "authenticate with a blank credential."
  end

  def message({:provider_failed, name, provider, detail}) do
    "the #{provider} provider failed resolving #{inspect(name)}: #{detail}"
  end
end
